silk-core/src/main/kotlin/net/silkmc/silk/core/event/Event.kt

```diff
@@ -0,0 +1,136 @@
+package net.silkmc.silk.core.event
+
+import kotlinx.coroutines.*
+import kotlinx.coroutines.flow.FlowCollector
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.SharedFlow
+import net.silkmc.silk.core.annotations.ExperimentalSilkApi
+import net.silkmc.silk.core.annotations.InternalSilkApi
+import net.silkmc.silk.core.task.mcCoroutineDispatcher
+
+/**
+ * The base object for discovering events, check the extensions
+ * values of this object.
+ */
+@ExperimentalSilkApi
+object Events
+
+@ExperimentalSilkApi
+open class Event<T> {
+
+    @InternalSilkApi
+    val listeners: MutableList<(T) -> Unit> = ArrayList()
+
+    /**
+     * Listens to this event. The [callback] will always be called synchronously.
+     * This function is synchronized, so it may be called from any thread.
+     */
+    fun listen(callback: (event: T) -> Unit) {
+        synchronized(this) {
+            listeners += callback
+        }
+    }
+
+    /**
+     * Invokes this event. Calling this function will trigger all
+     * listeners and collectors.
+     */
+    open fun invoke(instance: T) {
+        synchronized(this) {
```

Comment:
It's not ideal, but the API should always be threadsafe and simply removing this would mean that the user has to be extra careful, which is not ideal. So this synchronized block should stay there *until* there is a better solution for a thread safe basic callback list.

Reply:
Yep totally agree, we would need a better solution that still allows usage from any thread